src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceGroupCompleter.cs

```diff
@@ -0,0 +1,142 @@
+    {
+        private static IDictionary<int, IList<String>> _resourceGroupNamesDictionary = new ConcurrentDictionary<int, IList<string>>();
+        private static readonly object _lock = new object();
+
+        protected static IList<String> ResourceGroupNames
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    IAzureContext context = AzureRmProfileProvider.Instance.Profile.DefaultContext;
+                    var contextHash = HashContext(context);
+                    if (!_resourceGroupNamesDictionary.ContainsKey(contextHash))
+                    {
+                        _resourceGroupNamesDictionary[contextHash] = new List<string>();
+                        try
+                        {
+                            var instance = AzureSession.Instance;
+                            var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            // Retrieve only the first page of ResourceGroups to display to the user
+                            var resourceGroups = client.ResourceGroups.ListAsync();
+                            if (resourceGroups.Wait(TimeSpan.FromSeconds(5)))
+                            {
+                                if (resourceGroups.Result != null)
+                                {
+                                    var resourceGroupList = resourceGroups.Result;
+                                    foreach (ResourceGroup resourceGroup in resourceGroupList)
+                                    {
+                                        _resourceGroupNamesDictionary[contextHash].Add(resourceGroup.Name);
+                                    }
+                                }
+#if DEBUG
+                                else
+                                {
+                                    throw new Exception("Result from client.ResourceGroups is null");
+                                }
+#endif
+                            }
+#if DEBUG
+                            else
+                            {
+                                throw new Exception("client.ResourceGroups call timed out");
+                            }
+#endif
+                        }
+
+                        catch (Exception ex)
+                        {
+#if DEBUG
+                            throw ex;
+#endif
+                        }
+                    }
+
+                    return _resourceGroupNamesDictionary[contextHash];
```

Comment:
so, if we haven't previously cached a value and the call actually times out in production, this will fall through to here, which will null ref, I think.  We probably need to store the result in a local variable and return that, with the default value beign returned if the call to list resource groups times out.

Reply:
Discussed offline - no changes necessary.